include/mbgl/gl/gl.hpp

```diff
@@ -36,12 +38,31 @@ struct Error : std::runtime_error {
 
 void checkError(const char *cmd, const char *file, int line);
 
-} // namespace gl
-} // namespace mbgl
-
 #ifndef NDEBUG
 #define MBGL_CHECK_ERROR(cmd) ([&]() { struct __MBGL_C_E { ~__MBGL_C_E() { ::mbgl::gl::checkError(#cmd, __FILE__, __LINE__); } } __MBGL_C_E; return cmd; }())
 #else
 #define MBGL_CHECK_ERROR(cmd) (cmd)
 #endif
 
+template <typename T> struct AttributeType;
+
+template <> struct AttributeType<int8_t>   : std::integral_constant<GLenum, GL_BYTE> {};
+template <> struct AttributeType<uint8_t>  : std::integral_constant<GLenum, GL_UNSIGNED_BYTE> {};
+template <> struct AttributeType<int16_t>  : std::integral_constant<GLenum, GL_SHORT> {};
+template <> struct AttributeType<uint16_t> : std::integral_constant<GLenum, GL_UNSIGNED_SHORT> {};
+template <> struct AttributeType<int32_t>  : std::integral_constant<GLenum, GL_INT> {};
+template <> struct AttributeType<uint32_t> : std::integral_constant<GLenum, GL_UNSIGNED_INT> {};
+template <> struct AttributeType<float>    : std::integral_constant<GLenum, GL_FLOAT> {};
+
+template <class V, class E, std::size_t N>
+void bindVertexAttribute(AttributeLocation location, E (V::*)[N], const int8_t* offset, std::size_t memberOffset) {
+    MBGL_CHECK_ERROR(glEnableVertexAttribArray(location));
+    MBGL_CHECK_ERROR(glVertexAttribPointer(location, N, AttributeType<E>::value, false, sizeof(V), offset + memberOffset));
+}
+
+// This has to be a macro because it uses the offsetof macro, which is the only legal way to get a member offset.
+#define MBGL_BIND_VERTEX_ATTRIBUTE(VertexType, member, offset) \
+    ::mbgl::gl::bindVertexAttribute(Shader::member, &VertexType::member, offset, offsetof(VertexType, member))
```

Comment:
We should be using `std::is_standard_layout` to ensure that we can safely use `offsetof`. We should also assert that all values we get from `offsetof` are [multiples of 4](https://developer.apple.com/library/content/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/TechniquesforWorkingwithVertexData/TechniquesforWorkingwithVertexData.html#//apple_ref/doc/uid/TP40008793-CH107-SW7). 

It also assumes that the shader and the vertex type have identically named member names. I feel a little bit uneasy about that, but can't quite put my finger on it.